regex/src/main/scala/regex/Regex.scala

```diff
@@ -1,35 +1,70 @@
 package regex
 
 /** *****************************************************************************
-  * Regular Languages
-  *
-  * data structure definitions for regular languages
-  */
+  * Regular Languages
+  *
+  * data structure definitions for regular languages
+  */
 
-// Add your definitions here
+trait RegularLanguage
 
-/** *****************************************************************************
-  * Derivatives
-  *
-  * Fill in the function definitions below
-  */
+case object Empty extends RegularLanguage
+
+case object Epsilon extends RegularLanguage
+
+case class Character(c: Char) extends RegularLanguage
+
+case class Union(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Concat(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Star(l: RegularLanguage) extends RegularLanguage
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = lang match {
+  case Concat(Epsilon, sublang) => simplify(sublang)
+  case Concat(sublang, Epsilon) => simplify(sublang)
+  case Concat(Empty, sublang) => simplify(Empty)
+  case Concat(sublang, Empty) => simplify(Empty)
+  case Concat(sublang1, sublang2) => Concat(simplify(sublang1), simplify(sublang2))
+  case Union(Empty, sublang) => simplify(sublang)
+  case Union(sublang, Empty) => simplify(sublang)
+  case Union(sublang1, sublang2) => Union(simplify(sublang1), simplify(sublang2))
+  case Star(Epsilon) => Epsilon
+  case Star(Empty) => Empty
+  case Star(sublang) => Star(simplify(sublang))
+  case _ => lang
+}
+
 
 /** A language is nullable if it contains ε */
-def nullable(lang: RegularLanguage): Boolean = ???
+def nullable(lang: RegularLanguage): Boolean = lang match {
+  case Empty => false
+  case Epsilon => true
+  case Character(_) => false
+  case Union(l1, l2) => nullable(l1) || nullable(l2)
+  case Concat(l1, l2) => nullable(l1) && nullable(l2)
+  case Star(_) => true
+}
 
 /** Computes the derivative of a language, with respect to a character */
-def derivative(l: RegularLanguage)(c: Char): RegularLanguage = ???
+def derivative(l: RegularLanguage)(c: Char): RegularLanguage = l match {
+  case Empty | Epsilon => Empty
```

Comment:
Everything in this function seemed similar to my code, with the exception of this or case, I did these on separate lines but I like this approach too. Not sure if this could have been done in the nullable function as well but that would be interesting to explore